{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4b8a1994",
   "metadata": {},
   "source": [
    "# make conda environment & install packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95640a55",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "conda deactivate; conda remove -n WGCNA --all -y\n",
    "mamba create -n WGCNA \\\n",
    "    -c conda-forge -c bioconda -c nodefaults --strict-channel-priority \\\n",
    "    r-base libxml2 seaborn-base scnic r-car  r-mice r-igraph r-leidenbase r-xml2 r-xml r-rcurl r-restfulr bioconductor-rtracklayer bioconductor-rgraphviz bioconductor-gostats r-hdf5r xz julia -y\n",
    "\n",
    "mamba activate WGCNA\n",
    "R\n",
    "install.packages(c(\"remotes\",\"BiocManager\",\"hdf5r\"), repos=\"https://ftp.yz.yamagata-u.ac.jp/pub/cran\")\n",
    "BiocManager::install(c(\"WGCNA\",\"GenomicRanges\",\"GeneOverlap\",\"UCell\",\"glmGamPoi\",\"clusterProfiler\",\"org.Hs.eg.db\",\"AnnotationDbi\",\"zellkonverter\"), ask=F)\n",
    "BiocManager::install(\"SGCP\", ask=FALSE, lib = .libPaths()[1])\n",
    "remotes::install_github(\"satijalab/seurat\", \"seurat5\", quiet = TRUE)\n",
    "remotes::install_github(\"smorabit/hdWGCNA\", ref=\"dev\", quiet = TRUE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "632a304f",
   "metadata": {},
   "source": [
    "# DL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ca4c95",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "API=\"https://storage.googleapis.com/storage/v1/b/adult-gtex/o\"\n",
    "TMP_URLS=\"urllist.txt\"\n",
    ": > \"$TMP_URLS\"\n",
    "\n",
    "# prefix をクエリ用にエンコード（jqを使って安全に）\n",
    "PREFIX=\"$(jq -rn --arg p \"bulk-gex/v10/rna-seq/tpms-by-tissue/\" '$p|@uri')\"\n",
    "\n",
    "token=\"\"\n",
    "\n",
    "# === 一覧取得（ページング対応）===\n",
    "while :; do\n",
    "    url=\"${API}?prefix=${PREFIX}&fields=items(name),nextPageToken\"\n",
    "    [[ -n \"${token}\" ]] && url=\"${url}&pageToken=${token}\"\n",
    "\n",
    "    json=\"$(curl -fsS \"$url\")\"\n",
    "\n",
    "    # .gz のみ抽出し、各パス要素をURLエンコードしつつ、/ はそのまま残す\n",
    "    # （スペース等は %20 になるが、スラッシュは %2F にならない）\n",
    "    echo \"$json\" \\\n",
    "    | jq -r '\n",
    "        .items[]?.name\n",
    "        | select(endswith(\".gz\"))\n",
    "        | (split(\"/\") | map(@uri) | join(\"/\"))\n",
    "        | \"https://storage.googleapis.com/adult-gtex/\\(.)\"\n",
    "        ' >> \"$TMP_URLS\"\n",
    "\n",
    "    token=\"$(echo \"$json\" | jq -r '.nextPageToken // empty')\"\n",
    "    [[ -z \"$token\" ]] && break\n",
    "done\n",
    "\n",
    "wget -c -nc -i \"$TMP_URLS\"\n",
    "gunzip -n *.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce90f61b",
   "metadata": {},
   "source": [
    "# データ前処理"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba3af7f4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "options(future.globals.maxSize = 100 * 1024^3)\n",
    "Pkgs <- c(\"SGCP\", \"org.Hs.eg.db\", \"AnnotationDbi\")\n",
    "for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "\n",
    "gct_files <- c(\n",
    "        \"adipose_subcutaneous\", \"adipose_visceral_omentum\", \"artery_tibial\", \"breast_mammary_tissue\", \"cells_cultured_fibroblasts\",\n",
    "        \"esophagus_mucosa\", \"esophagus_muscularis\", \"lung\",\"muscle_skeletal\",\"nerve_tibial\",\n",
    "        \"skin_not_sun_exposed_suprapubic\", \"skin_sun_exposed_lower_leg\", \"thyroid\", \"whole_blood\"\n",
    ")\n",
    "LoadData(\"data/GTEx_Analysis_v10_RNASeQCv2.4.2_gene_tpm.gct\")\n",
    "gct_file <- \"kidney_medulla\"\n",
    "\n",
    "LoadData <- function(fn) {\n",
    "        data <- read.csv(paste0(\"data/gene_tpm_v10_\",fn,\".gct\"), sep=\"\\t\", header=TRUE, skip=2)\n",
    "        data <- data[!duplicated(data[, 2]) & !duplicated(data[, 2], fromLast = TRUE), ]\n",
    "        genes <- AnnotationDbi::select(org.Hs.eg.db, keys=data$Description, columns=c(\"ENTREZID\"), keytype=\"SYMBOL\")\n",
    "        genes <- genes[(!is.na(genes$SYMBOL)) & (!is.na(genes$ENTREZID)), ]\n",
    "        genes <- genes[(!duplicated(genes$SYMBOL)) & (!duplicated(genes$ENTREZID)),]\n",
    "        names(data)[2] <- \"SYMBOL\"\n",
    "        data <- merge(genes, data, by=\"SYMBOL\")\n",
    "        genes <- data[, 1:2]\n",
    "        data <- data[, -c(1:3)]\n",
    "        vars <- apply(data, 1, var)\n",
    "        zeroInd <- which(vars == 0)\n",
    "        if(length(zeroInd) != 0) {\n",
    "                data <- data[-zeroInd, ]\n",
    "                genes <- genes[-zeroInd, ]\n",
    "        }\n",
    "        xx <- as.list(org.Hs.egGO[genes$ENTREZID])\n",
    "        haveGO <- sapply(xx, function(x) {if (length(x) == 1 && is.na(x)) FALSE else TRUE })\n",
    "        numNoGO <- sum(!haveGO)\n",
    "        if(numNoGO != 0){\n",
    "                data <- data[haveGO, ]\n",
    "                genes <- genes[haveGO, ]\n",
    "        }\n",
    "        rownames(data) <- genes$SYMBOL\n",
    "        write.table(data, paste0(\"tmp/\",fn,\".tsv\"), sep='\\t', col.names=NA, quote=FALSE)\n",
    "        # write.table(data, paste0(\"tmp/all.tsv\"), sep='\\t', col.names=NA, quote=FALSE)\n",
    "        rownames(data) <- genes$ENTREZID\n",
    "        saveRDS(list(expData=data, geneID=genes), paste0(\"tmp/\",fn,\".rds\"))\n",
    "        # saveRDS(list(expData=data, geneID=genes), paste0(\"tmp/all.rds\"))\n",
    "}\n",
    "for (gct_file in gct_files) {\n",
    "        LoadData(gct_file)\n",
    "        print(gct_file)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89287d93",
   "metadata": {},
   "source": [
    "# SGCP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1846f2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "options(future.globals.maxSize = 100 * 1024^3)\n",
    "Pkgs <- c(\"SGCP\", \"org.Hs.eg.db\", \"AnnotationDbi\")\n",
    "for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "\n",
    "# ARGS <- commandArgs(trailingOnly = T)\n",
    "# I <- as.integer(ARGS[1])\n",
    "\n",
    "# 安定性\n",
    "## 観測単位\n",
    "gct_files <- c(\"adipose_subcutaneous\",\"muscle_skeletal\",\"skin_sun_exposed_lower_leg\",\"whole_blood\",\"all\")\n",
    "gct_file <- gct_files[I]\n",
    "for (gct_file in gct_files) {\n",
    "        data <- readRDS(paste0(\"tmp/\",gct_file,\".rds\"))\n",
    "        sgcp <- ezSGCP(expData=as.matrix(data$expData), geneID=data$geneID$ENTREZID, annotation_db=\"org.Hs.eg.db\", eff.egs=FALSE , saveOrig=FALSE, sil=TRUE, hm=NULL)\n",
    "\n",
    "        names(data$geneID) <- c(\"Symbol\",\"geneID\")\n",
    "        res <- merge(data$geneID, sgcp$clusterLabels, by=\"geneID\")\n",
    "  \n",
    "        write.table(res[,c(2,4)], paste0(\"Result/SGCP/\",gct_file,\".tsv\"), sep='\\t', row.names=FALSE)\n",
    "}\n",
    "\n",
    "## 変数\n",
    "Vars <- c(1000, 5000, 10000)\n",
    "data <- readRDS(paste0(\"tmp/all.rds\"))\n",
    "for (v in Vars) {\n",
    "        sgcp <- ezSGCP(expData=as.matrix(data$expData[1:v,]), geneID=data$geneID$ENTREZID[1:v], annotation_db=\"org.Hs.eg.db\", eff.egs=FALSE , saveOrig=FALSE, sil=TRUE, hm=NULL)\n",
    "\n",
    "        names(data$geneID) <- c(\"Symbol\",\"geneID\")\n",
    "        res <- merge(data$geneID, sgcp$clusterLabels, by=\"geneID\")\n",
    "  \n",
    "        write.table(res[,c(2,4)], paste0(\"Result/SGCP/\",v,\".tsv\"), sep='\\t', row.names=FALSE)\n",
    "}\n",
    "\n",
    "\n",
    "# 実行時間測定\n",
    "options(warn=-1)\n",
    "options(future.globals.maxSize = 100 * 1024^3)\n",
    "Pkgs <- c(\"SGCP\", \"org.Hs.eg.db\", \"AnnotationDbi\")\n",
    "for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "data <- readRDS(\"tmp/adipose_subcutaneous.rds\")\n",
    "params_obs <- c(100, 200, 400)\n",
    "params_obs <- c(400)\n",
    "params_var <- c(500, 1000, 5000, 10000)\n",
    "work <- function(buf) {\n",
    "        set.seed(42)\n",
    "        sgcp <- ezSGCP(expData=buf$expData, geneID=buf$geneID, annotation_db=\"org.Hs.eg.db\", eff.egs=FALSE , saveOrig=FALSE, sil=TRUE, hm=NULL)\n",
    "        # names(data$geneID) <- c(\"Symbol\",\"geneID\")\n",
    "        # res <- merge(data$geneID, sgcp$clusterLabels, by=\"geneID\")\n",
    "        # names(res)[4] <- \"SGCP\"\n",
    "  \n",
    "        # write.table(res[,c(2,4)], paste0(\"Result/SGCP/\",buf$fn,\".tsv\"), sep='\\t', row.names=FALSE, col.names=NA)\n",
    "        return(NULL)\n",
    "}\n",
    "## 観測単位数\n",
    "results <- do.call(\n",
    "        rbind,\n",
    "        lapply(params_obs, function(p) {\n",
    "                trials <- replicate(10, {\n",
    "                    buf <- list(expData=as.matrix(data$expData[1:1000,1:p]), geneID=data$geneID$ENTREZID[1:1000])\n",
    "                        elapsed <- as.numeric(system.time(work(buf))[\"elapsed\"])\n",
    "                        data.frame(param=p, elapsed=elapsed)\n",
    "                }, simplify = FALSE)\n",
    "                do.call(rbind, trials)\n",
    "        })\n",
    ")\n",
    "buf <- readRDS(\"tmp_SGCP_obs.rds\")\n",
    "results <- rbind(buf, results)\n",
    "saveRDS(results, file=\"tmp_SGCP_obs.rds\")\n",
    "## 変数数\n",
    "# params_var <- c(500, 1000, 5000, 10000)\n",
    "params_var <- c(5000)\n",
    "results <- do.call(\n",
    "    rbind,\n",
    "    lapply(params_var, function(p) {\n",
    "            trials <- replicate(10, {\n",
    "                    buf <- list(expData=as.matrix(data$expData[1:p,1:100]), geneID=data$geneID$ENTREZID[1:p])\n",
    "                    elapsed <- as.numeric(system.time(work(buf))[\"elapsed\"])\n",
    "                    data.frame(param=p, elapsed=elapsed)\n",
    "            }, simplify = FALSE)\n",
    "            do.call(rbind, trials)\n",
    "    })\n",
    ")\n",
    "results <- do.call(\n",
    "        rbind,\n",
    "        lapply(params_var, function(p) {\n",
    "                trials <- list()\n",
    "                success <- 0\n",
    "                while (success < 10) {\n",
    "                        buf <- list(expData=as.matrix(data$expData[1:p,1:100]), geneID=data$geneID$ENTREZID[1:p])\n",
    "                        elapsed <- tryCatch({\n",
    "                                as.numeric(system.time(work(buf))[\"elapsed\"])\n",
    "                        }, error = function(e) {\n",
    "                                NULL\n",
    "                        })\n",
    "                        if (!is.null(elapsed)) {\n",
    "                                success <- success + 1\n",
    "                                trials[[length(trials) + 1]] <- data.frame(param=p, elapsed=elapsed)\n",
    "                        }\n",
    "                }\n",
    "                do.call(rbind, trials)\n",
    "        })\n",
    ")\n",
    "\n",
    "buf <- readRDS(\"tmp_SGCP_var.rds\")\n",
    "results <- rbind(buf, results)\n",
    "saveRDS(results, file=\"tmp_SGCP_var.rds\")\n",
    "\n",
    "write.table(results, \"Result/SGCP/time_obs.tsv\", sep='\\t', row.names=FALSE, quote=FALSE)\n",
    "write.table(results, \"Result/SGCP/time_var.tsv\", sep='\\t', row.names=FALSE, quote=FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86bd3e92",
   "metadata": {},
   "source": [
    "# WGCNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd2efe4b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "Pkgs <- c(\"WGCNA\", \"org.Hs.eg.db\", \"AnnotationDbi\")\n",
    "for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "\n",
    "options(stringsAsFactors = FALSE)\n",
    "# enableWGCNAThreads()\n",
    "\n",
    "# 安定性\n",
    "## 観測単位\n",
    "gct_files <- c(\"adipose_subcutaneous\",\"muscle_skeletal\",\"skin_sun_exposed_lower_leg\",\"whole_blood\",\"all\")\n",
    "ExpMats <- list()\n",
    "Beta <- c(1:50)\n",
    "for (gct_file in gct_files){\n",
    "    data <- readRDS(paste0(\"tmp/\",gct_file,\".rds\"))\n",
    "    ExpMats[[gct_file]] <- t(as.matrix(data$expData))\n",
    "\n",
    "    Sft <- pickSoftThreshold(ExpMats[[gct_file]], powerVector=Beta)\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_SI_\",gct_file,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        print(\n",
    "            plot(\n",
    "                    Sft$fitIndices[,1], -sign(Sft$fitIndices[,3])*Sft$fitIndices[,2],\n",
    "                    xlab=\"Soft Threshold (power)\", ylab=\"Scale Free Topology Model Fit,signed R^2\", type=\"n\",\n",
    "                    main=paste(\"Scale independence\")\n",
    "                ) +\n",
    "            text(\n",
    "                    Sft$fitIndices[,1], -sign(Sft$fitIndices[,3])*Sft$fitIndices[,2],\n",
    "                    labels=Beta, cex=0.9, col=\"red\"\n",
    "                ) +\n",
    "            abline(h=0.90,col=\"red\")\n",
    "        )\n",
    "    dev.off()\n",
    "\n",
    "    print(gct_file)\n",
    "}\n",
    "saveRDS(ExpMats, file=\"tmp_WGCNA/ExpMats.rds\")\n",
    "# ExpMats <- readRDS(\"tmp_WGCNA/ExpMats.rds\")\n",
    "Betas <- list(adipose_subcutaneous=8,muscle_skeletal=11,skin_sun_exposed_lower_leg=11,whole_blood=32,all=4)\n",
    "Thres <- list(adipose_subcutaneous=0.34,muscle_skeletal=0.34,skin_sun_exposed_lower_leg=0.32,whole_blood=0.3,all=0.3)\n",
    "Adj <- list()\n",
    "for (gct_file in gct_files){\n",
    "    Adj[[gct_file]] <- adjacency(ExpMats[[gct_file]], power=Betas[[gct_file]])\n",
    "    # Adj[[gct_file]] <-  readRDS(paste0(\"tmp_WGCNA/\",gct_file,\".rds\"))\n",
    "    k <- as.vector(apply(Adj[[gct_file]], 2, sum, na.rm=T))\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_ScaleFreePlot_\",gct_file,\".pdf\"), width=10/2.54, height=10/2.54)\n",
    "        scaleFreePlot(k, main=\"Check scale free topology\\n\")\n",
    "    dev.off()\n",
    "    # saveRDS(Adj[[gct_file]], file=paste0(\"tmp_WGCNA/\",gct_file,\".rds\"))\n",
    "\n",
    "    TOM <- TOMsimilarity(Adj[[gct_file]])\n",
    "    DissTOM <- 1 - TOM\n",
    "    GeneTree = hclust(as.dist(DissTOM), method=\"average\")\n",
    "    DynamicMods <- cutreeDynamic(\n",
    "                                    dendro=GeneTree, \n",
    "                                    distM=DissTOM,\n",
    "                                    deepSplit=4, \n",
    "                                    pamStage=TRUE,\n",
    "                                    pamRespectsDendro=TRUE,\n",
    "                                    minClusterSize=30\n",
    "                                )\n",
    "    DynamicColors <- labels2colors(DynamicMods)\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_Dendrogram_1_\",gct_file,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        print(\n",
    "            plotDendroAndColors(\n",
    "                                    dendro=GeneTree, \n",
    "                                    colors=DynamicColors, \n",
    "                                    groupLabels=\"Dynamic Tree Cut\",\n",
    "                                    dendroLabels=FALSE, hang=0.03,\n",
    "                                    addGuide=TRUE, guideHang=0.05,\n",
    "                                    main=\"Gene dendrogram and module colors\"\n",
    "                                )\n",
    "        )\n",
    "    dev.off()\n",
    "\n",
    "    MEList <- moduleEigengenes(ExpMats[[gct_file]], colors=DynamicColors)\n",
    "    MEs <- MEList$eigengenes\n",
    "    MEDiss <- 1 - cor(MEs)\n",
    "    METree <- hclust(as.dist(MEDiss), method=\"average\")\n",
    "    MEDissThres <- Thres[[gct_file]]\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_ClusteringModule_\",gct_file,\"_\",MEDissThres,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        print(\n",
    "            plot(\n",
    "                    METree,\n",
    "                    main=\"Clustering of module eigengenes\",\n",
    "                    xlab = \"\", sub = \"\"\n",
    "                ) +\n",
    "            abline(h=MEDissThres, col=\"red\")\n",
    "        )\n",
    "    dev.off()\n",
    "}\n",
    "for (gct_file in gct_files){\n",
    "    Merge <- mergeCloseModules(ExpMats[[gct_file]], DynamicColors, cutHeight=MEDissThres, verbose=3)\n",
    "    MergedColours <- Merge$colors\n",
    "    MergedMEs <- Merge$newMEs\n",
    "    png(paste0(\"Result/Fig/WGCNA_Dendrogram_2.png\"), width=960, height=480)\n",
    "        plotDendroAndColors(\n",
    "                                dendro=GeneTree,\n",
    "                                colors=cbind(DynamicColors, MergedColours),\n",
    "                                groupLabels=c(\"Dynamic Tree Cut\", \"Merged dynamic\"),\n",
    "                                dendroLabels=FALSE, hang=0.03,\n",
    "                                addGuide=TRUE, guideHang=0.05\n",
    "                            )\n",
    "    dev.off()\n",
    "\n",
    "    ColourList <- data.frame(colnames(ExpMats[[gct_file]]), MergedColours)\n",
    "    colnames(ColourList) <- c(\"Symbol\", \"WGCNA\")\n",
    "    write.table(ColourList, \"Result/WGCNA.tsv\", sep='\\t', row.names=F)\n",
    "}\n",
    "## 変数\n",
    "Vars <- c(1000, 5000, 10000)\n",
    "Data_All <- readRDS(paste0(\"tmp/all.rds\"))\n",
    "ExpMats <- list()\n",
    "Beta <- c(1:50)\n",
    "for (v in Vars){\n",
    "    ExpMats[[as.character(v)]] <- t(as.matrix(Data_All$expData[1:v,]))\n",
    "\n",
    "    Sft <- pickSoftThreshold(ExpMats[[as.character(v)]], powerVector=Beta)\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_SI_\",v,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        print(\n",
    "            plot(\n",
    "                    Sft$fitIndices[,1], -sign(Sft$fitIndices[,3])*Sft$fitIndices[,2],\n",
    "                    xlab=\"Soft Threshold (power)\", ylab=\"Scale Free Topology Model Fit,signed R^2\", type=\"n\",\n",
    "                    main=paste(\"Scale independence\")\n",
    "                ) +\n",
    "            text(\n",
    "                    Sft$fitIndices[,1], -sign(Sft$fitIndices[,3])*Sft$fitIndices[,2],\n",
    "                    labels=Beta, cex=0.9, col=\"red\"\n",
    "                ) +\n",
    "            abline(h=0.90,col=\"red\")\n",
    "        )\n",
    "    dev.off()\n",
    "\n",
    "    print(v)\n",
    "}\n",
    "saveRDS(ExpMats, file=\"tmp_WGCNA/ExpMatsV.rds\")\n",
    "Betas <- list(\"1000\"=47,\"5000\"=21,\"10000\"=15)\n",
    "Thres <- list(\"1000\"=0.3,\"5000\"=0.3,\"10000\"=0.35)\n",
    "for (v in Vars){\n",
    "    Adj <- adjacency(ExpMats[[as.character(v)]], power=Betas[[as.character(v)]])\n",
    "    k <- as.vector(apply(Adj, 2, sum, na.rm=T))\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_ScaleFreePlot_\",v,\".pdf\"), width=10/2.54, height=10/2.54)\n",
    "        print(scaleFreePlot(k, main=\"Check scale free topology\\n\"))\n",
    "    dev.off()\n",
    "\n",
    "    TOM <- TOMsimilarity(Adj)\n",
    "    DissTOM <- 1 - TOM\n",
    "    GeneTree = hclust(as.dist(DissTOM), method=\"average\")\n",
    "    DynamicMods <- cutreeDynamic(\n",
    "                                    dendro=GeneTree, \n",
    "                                    distM=DissTOM,\n",
    "                                    deepSplit=4, \n",
    "                                    pamStage=TRUE,\n",
    "                                    pamRespectsDendro=TRUE,\n",
    "                                    minClusterSize=30\n",
    "                                )\n",
    "    DynamicColors <- labels2colors(DynamicMods)\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_Dendrogram_1_\",v,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        print(\n",
    "            plotDendroAndColors(\n",
    "                                    dendro=GeneTree, \n",
    "                                    colors=DynamicColors, \n",
    "                                    groupLabels=\"Dynamic Tree Cut\",\n",
    "                                    dendroLabels=FALSE, hang=0.03,\n",
    "                                    addGuide=TRUE, guideHang=0.05,\n",
    "                                    main=\"Gene dendrogram and module colors\"\n",
    "                                )\n",
    "        )\n",
    "    dev.off()\n",
    "\n",
    "    MEList <- moduleEigengenes(ExpMats[[as.character(v)]], colors=DynamicColors)\n",
    "    MEs <- MEList$eigengenes\n",
    "    MEDiss <- 1 - cor(MEs)\n",
    "    METree <- hclust(as.dist(MEDiss), method=\"average\")\n",
    "    MEDissThres <- Thres[[as.character(v)]]\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_ClusteringModule_\",v,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        print(\n",
    "            plot(\n",
    "                    METree,\n",
    "                    main=\"Clustering of module eigengenes\",\n",
    "                    xlab = \"\", sub = \"\"\n",
    "                ) +\n",
    "            abline(h=MEDissThres, col=\"red\")\n",
    "        )\n",
    "    dev.off()\n",
    "\n",
    "    Merge <- mergeCloseModules(ExpMats[[as.character(v)]], DynamicColors, cutHeight=MEDissThres, verbose=3)\n",
    "    MergedColours <- Merge$colors\n",
    "    MergedMEs <- Merge$newMEs\n",
    "    pdf(paste0(\"Result/Fig/WGCNA_Dendrogram_2_\",v,\".pdf\"), width=32.4/2.54, height=20/2.54)\n",
    "        plotDendroAndColors(\n",
    "                                dendro=GeneTree,\n",
    "                                colors=cbind(DynamicColors, MergedColours),\n",
    "                                groupLabels=c(\"Dynamic Tree Cut\", \"Merged dynamic\"),\n",
    "                                dendroLabels=FALSE, hang=0.03,\n",
    "                                addGuide=TRUE, guideHang=0.05\n",
    "                            )\n",
    "    dev.off()\n",
    "\n",
    "    ColourList <- data.frame(colnames(ExpMats[[as.character(v)]]), MergedColours)\n",
    "    colnames(ColourList) <- c(\"Symbol\", \"Colour\")\n",
    "    write.table(ColourList, paste0(\"Result/WGCNA/\",v,\".tsv\"), sep='\\t', row.names=F)\n",
    "}\n",
    "\n",
    "# 実行時間測定\n",
    "options(warn=-1)\n",
    "Pkgs <- c(\"WGCNA\", \"org.Hs.eg.db\", \"AnnotationDbi\")\n",
    "for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "data <- readRDS(\"tmp/adipose_subcutaneous.rds\")\n",
    "params_var <- c(500, 1000, 5000, 10000)\n",
    "params_obs <- c(100, 200, 400)\n",
    "work <- function(buf) {\n",
    "    set.seed(42)\n",
    "    fit <- pickSoftThreshold(buf, powerVector=1:50)\n",
    "    Adj <- adjacency(buf, power=10);\n",
    "    TOM <- TOMsimilarity(Adj)\n",
    "    DissTOM <- 1 - TOM\n",
    "    DynamicMods <- cutreeDynamic(\n",
    "                                    dendro=hclust(as.dist(DissTOM), method=\"average\"), \n",
    "                                    distM=DissTOM,\n",
    "                                    deepSplit=4, \n",
    "                                    pamStage=TRUE,\n",
    "                                    pamRespectsDendro=TRUE,\n",
    "                                    minClusterSize=30\n",
    "                                )\n",
    "    Merge <- mergeCloseModules(buf, labels2colors(DynamicMods), cutHeight=0.3)\n",
    "}\n",
    "### 観測単位数\n",
    "results <- do.call(\n",
    "    rbind,\n",
    "    lapply(params_obs, function(p) {\n",
    "            trials <- replicate(10, {\n",
    "                    buf <- t(as.matrix(data$expData[1:1000, 1:p]))\n",
    "                    elapsed <- as.numeric(system.time(work(buf))[\"elapsed\"])\n",
    "                    data.frame(param=p, elapsed=elapsed)\n",
    "            }, simplify = FALSE)\n",
    "            do.call(rbind, trials)\n",
    "    })\n",
    ")\n",
    "write.table(results, \"Result/WGCNA/time_obs.tsv\", sep='\\t', row.names=FALSE, quote=FALSE)\n",
    "### 変数数\n",
    "results <- do.call(\n",
    "    rbind,\n",
    "    lapply(params_var, function(p) {\n",
    "            trials <- replicate(10, {\n",
    "                    buf <- t(as.matrix(data$expData[1:p, 1:100]))\n",
    "                    elapsed <- as.numeric(system.time(work(buf))[\"elapsed\"])\n",
    "                    data.frame(param=p, elapsed=elapsed)\n",
    "            }, simplify = FALSE)\n",
    "            do.call(rbind, trials)\n",
    "    })\n",
    ")\n",
    "write.table(results, \"Result/WGCNA/time_var.tsv\", sep='\\t', row.names=FALSE, quote=FALSE)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a64d771",
   "metadata": {},
   "source": [
    "# SGCRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef9659a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### SGCRNA_ALL #####\n",
    "using CSV, DataFrames\n",
    "using StatsBase\n",
    "using JLD2\n",
    "using SGCRNAs\n",
    "using CairoMakie\n",
    "\n",
    "function dot_plot(res, n, fn)\n",
    "    cols = [(mod(i-1, n) + 1) for i in 1:length(res)]\n",
    "    rows = [ceil(Int, i / n) for i in 1:length(res)]\n",
    "    rows = (maximum(rows)+1) .- rows\n",
    "    sizes_raw  = [v[1] for v in res]\n",
    "    colours_raw = [v[2] for v in res]\n",
    "\n",
    "    smin, smax = extrema(sizes_raw)\n",
    "    sizes_px = 6 .+ 24 .* (sizes_raw .- smin) ./ (smax - smin + eps())\n",
    "\n",
    "    f = Figure(size=(800, 650))\n",
    "    ax = Axis(\n",
    "        f[1, 1];\n",
    "        xticks=1:n, yticks=1:n,\n",
    "        aspect=1, title=\"\",\n",
    "        xgridvisible=false, ygridvisible=false\n",
    "    )\n",
    "    hidespines!(ax); hidedecorations!(ax, grid=false)\n",
    "    xlims!(ax, 0.5, n+0.5); ylims!(ax, 0.5, n+0.5)\n",
    "\n",
    "    scatter!(ax, cols, rows; markersize=sizes_px, color=colours_raw, colormap=:heat, colorrange=extrema(colours_raw))\n",
    "    text!(ax, cols .- 0.3, rows .- 0.3, text=string.(1:length(res)), align=(:right, :baseline), fontsize=10, color = :black)\n",
    "\n",
    "    vals = [0, 5, 10]\n",
    "    sizes_for_legend = 6 .+ 24 .* (vals .- smin) ./ (smax - smin + eps())\n",
    "    handles = [MarkerElement(marker=:circle, color=:black, strokecolor=:transparent, markersize=s) for s in sizes_for_legend]\n",
    "    Legend(f[1, 2], handles, string.(vals), \"size value\", framevisible=false)\n",
    "    Colorbar(f[1, 3], limits = (0, 1.0), colormap=:heat, label=\"Simpson index\")\n",
    "\n",
    "    save(fn, f)\n",
    "end\n",
    "\n",
    "# 安定性・再現性・特異性\n",
    "fl = [\"adipose_subcutaneous\",\"muscle_skeletal\",\"skin_sun_exposed_lower_leg\",\"whole_blood\",\"all\"]\n",
    "## 観測単位\n",
    "for f in fl\n",
    "    Data = CSV.read(\"tmp/\"*f*\".tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "    CorData, GradData = CGM(Data.Column1, Matrix(Data[:,2:end]));\n",
    "    clust, pos, edge_data = SpectralClustering(CorData, GradData);\n",
    "    clust = clust[1]\n",
    "    gene_list = names(edge_data)\n",
    "    clust_list = []\n",
    "    for j in 1:maximum(clust)\n",
    "        push!(clust_list, gene_list[clust .== j])\n",
    "    end\n",
    "    save_object(\"tmp_SGCRNA/\"*fn*\".jld2\", clust_list);\n",
    "end\n",
    "## 変数\n",
    "Data = CSV.read(\"tmp/all.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "for i in [1000, 5000, 10000]\n",
    "    CorData, GradData = CGM(Data.Column1[1:i], Matrix(Data[1:i,2:end]));\n",
    "    clust, pos, edge_data = SpectralClustering(CorData, GradData);\n",
    "    clust = clust[1]\n",
    "    gene_list = names(edge_data)\n",
    "    clust_list = []\n",
    "    for j in 1:maximum(clust)\n",
    "        push!(clust_list, gene_list[clust .== j])\n",
    "    end\n",
    "    save_object(\"tmp_SGCRNA/\"*string(fn)*\".jld2\", clust_list);\n",
    "end\n",
    "\n",
    "# 実行時間測定\n",
    "using CSV, DataFrames\n",
    "using StatsBase\n",
    "using JLD2\n",
    "using SGCRNAs\n",
    "Data = CSV.read(\"tmp/adipose_subcutaneous.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "params_obs = [100, 200, 400];\n",
    "params_var = [500, 1000, 5000, 10000];\n",
    "function f(buf_g, buf)\n",
    "    CorData, GradData = CGM(buf_g, buf);\n",
    "    clust, pos, edge_data = SpectralClustering(CorData, GradData);\n",
    "\n",
    "    return nothing\n",
    "end\n",
    "f(Data.Column1[1:100], Matrix(Data[1:100, 2:101]));\n",
    "## 観測単位数\n",
    "Result = DataFrame(param=Int[], elapsed=Float64[])\n",
    "for p in params_obs\n",
    "    buf_g = Data.Column1[1:1000]\n",
    "    buf = Matrix(Data[1:1000, 2:(p+1)])\n",
    "\n",
    "    times = [@elapsed f(buf_g, buf) for _ in 1:10]\n",
    "    append!(Result, DataFrame(param=fill(p, 10), elapsed=times))\n",
    "    println(p)\n",
    "end\n",
    "## 変数数\n",
    "Result = DataFrame(param=Int[], elapsed=Float64[]);\n",
    "for p in params_var\n",
    "    buf_g = Data.Column1[1:p]\n",
    "    buf = Matrix(Data[1:p, 2:101])\n",
    "\n",
    "    times = [@elapsed f(buf_g, buf) for _ in 1:10]\n",
    "    append!(Result, DataFrame(param=fill(p, 10), elapsed=times))\n",
    "end\n",
    "Result |> CSV.write(\"Result/SGCRNA/time_obs.tsv\", delim='\\t')\n",
    "Result |> CSV.write(\"Result/SGCRNA/time_var.tsv\", delim='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d17ceba5",
   "metadata": {},
   "source": [
    "# 実行時間"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3876bb2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "using CSV, DataFrames\n",
    "using Colors, CairoMakie\n",
    "\n",
    "Tools = [\"SGCRNA\", \"WGCNA\", \"SGCP\"]\n",
    "# load data\n",
    "Obs = CSV.read(\"Result/SGCRNA/time_obs.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "rename!(Obs, \"elapsed\" => \"SGCRNA\")\n",
    "for i in 2:length(Tools)\n",
    "    buf = CSV.read(\"Result/\"*Tools[i]*\"/time_obs.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "    rename!(buf, \"elapsed\" => Tools[i])\n",
    "    Obs = hcat(Obs, buf[:,[Tools[i]]])\n",
    "end\n",
    "Obs = stack(Obs, 2:ncol(Obs))\n",
    "\n",
    "Var = CSV.read(\"Result/SGCRNA/time_var.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "rename!(Var, \"elapsed\" => \"SGCRNA\")\n",
    "for i in 2:length(Tools)\n",
    "    if i == 2\n",
    "        buf = CSV.read(\"Result/\"*Tools[i]*\"/time_var.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "    else\n",
    "        buf = CSV.read(\"Result/\"*Tools[i]*\"/time_var_buf.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "        buf = vcat(buf, DataFrame(param=vcat(repeat([5000],10),repeat([10000],10)), elapsed=vcat(repeat([669600],10),repeat([669600],10))))\n",
    "    end\n",
    "    rename!(buf, \"elapsed\" => Tools[i])\n",
    "    Var = hcat(Var, buf[:,[Tools[i]]])\n",
    "end\n",
    "Var = stack(Var, 2:ncol(Var))\n",
    "\n",
    "ToolTypes = unique(Obs.variable)\n",
    "Palette = [colorant\"rgb(255,75,0)\", colorant\"rgb(3,175,122)\", colorant\"rgb(0,90,255)\", colorant\"rgb(77,196,255)\"]\n",
    "ColourMap = Dict(ToolTypes[i] => Palette[i] for i in eachindex(ToolTypes))\n",
    "fig = Figure(size=(1200, 600), fontsize=14, markersize=16)\n",
    "ax = []\n",
    "push!(ax, Axis(fig[1, 1]; xlabel=\"# of observation\", ylabel=\"elapsed time (s)\", title=\"\", xgridvisible=false, ygridvisible=false, yscale=log10, limits=(0,nothing,1,10000)))\n",
    "push!(ax, Axis(fig[1, 2]; xlabel=\"# of variable\", ylabel=\"\", title=\"\", xgridvisible=false, ygridvisible=false, yscale=log10, limits=(0,nothing,1,10000)))\n",
    "linkyaxes!(ax[1], ax[2])\n",
    "for t in ToolTypes\n",
    "    scatter!(ax[1], Obs.param[Obs.variable .== t], Obs.value[Obs.variable .== t]; color=ColourMap[t], label=string(t), strokewidth=0)\n",
    "    scatter!(ax[2], Var.param[Var.variable .== t], Var.value[Var.variable .== t]; color=ColourMap[t], label=string(t), strokewidth=0)\n",
    "end\n",
    "fig[1, 3] = Legend(fig, ax[1]; framevisible=false)\n",
    "save(\"Result/Fig/elapsed.pdf\", fig)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2ff23d",
   "metadata": {},
   "source": [
    "# 安定性・再現性・特異性"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cc919b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SGCRNA\n",
    "using CSV, DataFrames\n",
    "using JLD2\n",
    "using Colors, CairoMakie\n",
    "\n",
    "function load_data(t::String, f::String)\n",
    "    Data = CSV.read(\"Result/\"*t*\"/\"*f*\".tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "    Clust = sort(unique(Data[:,2]))\n",
    "    return ([Data.Symbol[Data[:,2] .== c] for c in Clust], Clust)\n",
    "end\n",
    "fl = [\n",
    "    \"adipose_subcutaneous\",\"muscle_skeletal\",\"skin_sun_exposed_lower_leg\",\"whole_blood\",\n",
    "    \"1000\", \"5000\", \"10000\"\n",
    "]\n",
    "Size = Dict(\"SGCRNA\" => (2000,750), \"WGCNA\" => (2000,500), \"SGCP\" => (2000,500))\n",
    "for t in [\"SGCRNA\",\"WGCNA\",\"SGCP\"]\n",
    "    ## Simpson係数\n",
    "    All = Set()\n",
    "    Src_Clust = []\n",
    "    if t != \"SGCRNA\"\n",
    "        All, Src_Clust = load_data(t, \"all\")\n",
    "        All = Set.(All)\n",
    "    else\n",
    "        All = Set.(load_object(\"tmp_SGCRNA/all.jld2\"))\n",
    "        Src_Clust = collect(1:length(All))\n",
    "    end\n",
    "    N_All = length(All)\n",
    "    Result = []\n",
    "    ClustName =[]\n",
    "    for f in fl\n",
    "        Buf = Set()\n",
    "        if t != \"SGCRNA\"\n",
    "            Buf, Clust = load_data(t, f)\n",
    "            Buf = Set.(Buf)\n",
    "            push!(ClustName, Clust)\n",
    "        else\n",
    "            Buf = Set.(load_object(\"tmp_SGCRNA/\"*f*\".jld2\"))\n",
    "            push!(ClustName, collect(1:length(Buf)))\n",
    "        end\n",
    "        N_Buf = length(Buf)\n",
    "        Result_Buf = zeros(N_Buf, N_All)\n",
    "        for i in 1:N_All\n",
    "            for j in 1:N_Buf\n",
    "                Result_Buf[j, i] = length(All[i] ∩ Buf[j]) / min(length(All[i]), length(Buf[j]))\n",
    "            end\n",
    "        end\n",
    "\n",
    "        push!(Result, Result_Buf)\n",
    "    end\n",
    "\n",
    "    fig = Figure(size=Size[t], fontsize=20)\n",
    "    ax = []\n",
    "    for i in 1:length(Result)\n",
    "        if t == \"SGCP\"\n",
    "            push!(ax, Axis(fig[1, i]; xlabel=fl[i], ylabel=\"\", title=\"\", xgridvisible=false, ygridvisible=false, xticks=(1:size(Result[i],1), string.(ClustName[i])), xticklabelrotation=π/4, yticklabelsvisible=false, yticksvisible=false))\n",
    "        elseif t == \"WGCNA\"\n",
    "            push!(ax, Axis(fig[1, i]; xlabel=fl[i], ylabel=\"\", title=\"\", xgridvisible=false, ygridvisible=false, xticklabelsvisible=false, xticksvisible=false, yticklabelsvisible=false, yticksvisible=false))\n",
    "        else\n",
    "            push!(ax, Axis(fig[1, i]; xlabel=fl[i], ylabel=\"\", title=\"\", xgridvisible=false, ygridvisible=false, xticklabelrotation=π/4, yticklabelsvisible=false, yticksvisible=false))\n",
    "        end\n",
    "        heatmap!(ax[i], Result[i], colormap=:heat)\n",
    "    end\n",
    "    ax[1].ylabel = \"All\"\n",
    "    ax[1].yticklabelsvisible = true\n",
    "    ax[1].yticksvisible = true\n",
    "    if t == \"SGCP\"\n",
    "        ax[1].yticks = (1:size(Result[1],2), string.(Src_Clust))\n",
    "    elseif t == \"WGCNA\"\n",
    "        ax[1].yticklabelsvisible = false\n",
    "        ax[1].yticksvisible = false\n",
    "    end\n",
    "\n",
    "    Colorbar(fig[:, end+1], limits=(0, 1), colormap=:heat)\n",
    "    save(\"Result/Fig/\"*t*\"_quality.pdf\", fig)\n",
    "end\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.9.0",
   "language": "julia",
   "name": "julia-1.9"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
