{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f027050",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "module load singularity\n",
    "singularity pull docker://quay.io/qiime2/amplicon:2024.5\n",
    "\n",
    "wget https://data.qiime2.org/classifiers/sklearn-1.4.2/silva/silva-138-99-nb-classifier.qza\n",
    "\n",
    "wget https://github.com/picrust/picrust2/archive/v2.6.2.tar.gz\n",
    "tar xzf  v2.6.2.tar.gz; rm -f *.gz\n",
    "cd picrust2-2.6.2\n",
    "conda env create -f picrust2-env.yaml; conda activate picrust2\n",
    "pip install --editable .\n",
    "export TMPDIR=${HOME}/tmp\n",
    "pytest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e46fbfed",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#PJM -L rscunit=cx\n",
    "#PJM -L rscgrp=cx-share\n",
    "#PJM -L elapse=168:00:00\n",
    "#PJM -L jobenv=singularity\n",
    "\n",
    "module load singularity\n",
    "\n",
    "ls data/* \\\n",
    "| awk -v P=\"${PWD}\" 'BEGIN { print \"sample-id,absolute-filepath,direction\" } \\\n",
    "    { \\\n",
    "        orig=$0; \\\n",
    "        gsub(/data\\//, \"\", $1); \\\n",
    "        gsub(/_R[12].fastq.gz/, \"\", $1); \\\n",
    "        if(NR%2==0) { print $1 \",\" P \"/\" orig \",reverse\" } \\\n",
    "        else { print $1 \",\" P \"/\" orig \",forward\" } \\\n",
    "    }' \\\n",
    "> ManifestFile.csv\n",
    "\n",
    "# データ取込\n",
    "if [ ! -d ${PWD}/tmp ]; then mkdir -p ${PWD}/tmp; fi\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime tools import \\\n",
    "        --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "        --input-path ManifestFile.csv \\\n",
    "        --output-path tmp/demux.qza \\\n",
    "        --input-format PairedEndFastqManifestPhred33\n",
    "\n",
    "# 可視化\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime demux summarize   \\\n",
    "        --i-data tmp/demux.qza   \\\n",
    "        --o-visualization tmp/demux.qzv\n",
    "\n",
    "# PCRエラー修正\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime dada2 denoise-paired \\\n",
    "        --i-demultiplexed-seqs tmp/demux.qza \\\n",
    "        --p-trunc-len-f 250 \\\n",
    "        --p-trunc-len-r 250 \\\n",
    "        --p-n-threads 0 \\\n",
    "        --o-table tmp/table.qza \\\n",
    "        --o-denoising-stats tmp/stats.qza \\\n",
    "        --o-representative-sequences tmp/rep-seqs.qza \\\n",
    "        --verbose\n",
    "\n",
    "# 分類\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime feature-classifier classify-sklearn \\\n",
    "        --i-classifier ${HOME}/tool/silva-138-99-nb-classifier.qza \\\n",
    "        --i-reads tmp/rep-seqs.qza \\\n",
    "        --o-classification tmp/taxonomy.qza\n",
    "\n",
    "# 低リード数除去\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime feature-table filter-samples  \\\n",
    "        --i-table tmp/table.qza  \\\n",
    "        --p-min-frequency 1500 \\\n",
    "        --o-filtered-table tmp/selected_table.qza\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime feature-table filter-features  \\\n",
    "        --i-table tmp/selected_table.qza  \\\n",
    "        --p-min-frequency 1 \\\n",
    "        --o-filtered-table tmp/selected_table2.qza\n",
    "\n",
    "# 属レベルで統合\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime taxa collapse \\\n",
    "        --i-table tmp/selected_table2.qza \\\n",
    "        --i-taxonomy tmp/taxonomy.qza \\\n",
    "        --p-level 6 \\\n",
    "        --o-collapsed-table tmp/collapsed_table.qza\n",
    "\n",
    "# 補正\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime feature-table relative-frequency \\\n",
    "        --i-table tmp/collapsed_table.qza \\\n",
    "        --o-relative-frequency-table tmp/relfreq_table.qza\n",
    "\n",
    "# 出力\n",
    "if [ ! -d ${PWD}/Result ]; then mkdir -p ${PWD}/Result; fi\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime tools export \\\n",
    "        --input-path tmp/relfreq_table.qza \\\n",
    "        --output-path tmp\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    biom convert \\\n",
    "        -i tmp/feature-table.biom \\\n",
    "        -o Result/table.tsv \\\n",
    "        --to-tsv\n",
    "\n",
    "# picrust2\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime tools export \\\n",
    "        --input-path tmp/selected_table2.qza \\\n",
    "        --output-path tmp\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    biom convert \\\n",
    "        -i tmp/feature-table.biom \\\n",
    "        -o tmp/feature-table.tsv \\\n",
    "        --to-tsv\n",
    "sed -i '1d' tmp/feature-table.tsv\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/tool/amplicon_2024.5.sif \\\n",
    "    qiime tools export \\\n",
    "        --input-path tmp/rep-seqs.qza \\\n",
    "        --output-path tmp\n",
    "singularity exec \\\n",
    "    --bind $PWD:$PWD \\\n",
    "    ${HOME}/amplicon_2024.5.sif \\\n",
    "    qiime tools export \\\n",
    "        --input-path tmp/taxonomy.qza \\\n",
    "        --output-path tmp\n",
    "awk '$2 ~ /g__Faecalibacterium/ {print $1}' tmp/taxonomy.tsv > tmp/target_ids.txt\n",
    "awk '$2 ~ /g__Fusicatenibacter/ {print $1}' tmp/taxonomy.tsv > tmp/target_ids.txt\n",
    "awk '$2 ~ /g__\\[Eubacterium\\]_coprostanoligenes_group/ {print $1}' tmp/taxonomy.tsv > tmp/target_ids.txt\n",
    "grep -A 1 -f tmp/target_ids.txt tmp/dna-sequences.fasta | grep -v \"^--\" > tmp/selected.fasta\n",
    "head -n 1 tmp/feature-table.tsv > tmp/selected.tsv\n",
    "grep -F -f tmp/target_ids.txt tmp/feature-table.tsv >> tmp/selected.tsv\n",
    "\n",
    "picrust2_pipeline.py \\\n",
    "    -s tmp/selected.fasta \\\n",
    "    -i tmp/selected.tsv \\\n",
    "    -o Result/Eubacterium_coprostanoligenes_group \\\n",
    "    -p 16\n",
    "\n",
    "gunzip Result/Faecalibacterium/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz\n",
    "gunzip Result/Fusicatenibacter/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz\n",
    "gunzip Result/Eubacterium_coprostanoligenes_group/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz\n",
    "mv Result/Faecalibacterium/KO_metagenome_out/pred_metagenome_unstrat.tsv Result/Faecalibacterium/KO_metagenome_out/Faecalibacterium.tsv\n",
    "mv Result/Fusicatenibacter/KO_metagenome_out/pred_metagenome_unstrat.tsv Result/Fusicatenibacter/KO_metagenome_out/Fusicatenibacter.tsv\n",
    "mv Result/Eubacterium_coprostanoligenes_group/KO_metagenome_out/pred_metagenome_unstrat.tsv Result/Eubacterium_coprostanoligenes_group/KO_metagenome_out/Eubacterium_coprostanoligenes_group.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9443c68",
   "metadata": {
    "vscode": {
     "languageId": "julia"
    }
   },
   "outputs": [],
   "source": [
    "using CSV, DataFrames\n",
    "using JLD2\n",
    "using SGCRNAs\n",
    "\n",
    "\n",
    "Data = CSV.read(\"Result/table.tsv\", header=2, comment=\"\", delim='\\t', DataFrame);\n",
    "rename!(Data, 1=>:Taxon);\n",
    "# 属レベルでの分類データのみ取得\n",
    "Data = Data[occursin.(\"g__\",Data.Taxon) .& .!occursin.(\"g__uncultured\",Data.Taxon) .& .!occursin.(\"g__Incertae\",Data.Taxon), :];\n",
    "insertcols!(Data, 2, :Genus => string.(getindex.(split.(Data.Taxon, \";g__\"), 2)));\n",
    "\n",
    "CorData, GradData = CGM(Data.Genus, Matrix(Data[:,3:end]), fn=\"Result/\");\n",
    "# CorData = load_object(\"Result/_cor.jld2\");  GradData = load_object(\"Result/_grad.jld2\");\n",
    "Q = map(x -> sum(x)!=0.0, eachcol(CorData))\n",
    "CorData = CorData[Q,Q]\n",
    "GradData = GradData[Q,Q]\n",
    "\n",
    "# spectral clustering\n",
    "clust, pos, edge_data = SpectralClustering(CorData, GradData, tNodeNum=5, pcas=3, nNeighbors=1);\n",
    "save_object(\"Result/_scdata.jld2\", (clust, pos, edge_data));\n",
    "# clust, pos, edge_data = load_object(\"Result/_scdata.jld2\");\n",
    "\n",
    "# set parameter\n",
    "for i in 1:length(clust) println(maximum(clust[i])); end\n",
    "for i in 1:maximum(clust[1]) println(i, \"-\", sum(clust[1] .== i)); end\n",
    "\n",
    "d = 1; k = maximum(clust[d]);\n",
    "TaxonClust = DataFrame(Taxon=names(edge_data), Module=clust[d]);\n",
    "TaxonClust |> CSV.write(\"Result/TaxonCluster.tsv\", delim='\\t', writeheader=true)\n",
    "\n",
    "# draw network\n",
    "nw, new_pos, cnctdf, new_clust, score = SetNetwork(edge_data, clust[d], pos, il=collect(1:k));\n",
    "save_object(\"Result/_nwdata.jld2\", (nw, new_pos, cnctdf, new_clust, score, edge_data))\n",
    "# (nw, new_pos, cnctdf, new_clust, score, edge_data) = load_object(\"Result/_nwdata.jld2\"); k = maximum(new_clust);\n",
    "DrawNetwork(\"Result/MG_AllNetWork.png\", nw, new_pos, cnctdf, new_clust, k, node_scores=score, edge_mode=:ALL, edge_threshold=0.0, x_size=30, y_size=30)\n",
    "DrawNetwork(\"Result/MG_AllNetWork_labeled.png\", nw, new_pos, cnctdf, new_clust, k, node_scores=score, node_labels=sort(unique(vcat(cnctdf.e1,cnctdf.e2))), edge_mode=:ALL, edge_threshold=0.0, x_size=30, y_size=30)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "639f435a",
   "metadata": {},
   "source": [
    "# WGCNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2f76b8d",
   "metadata": {
    "vscode": {
     "languageId": "julia"
    }
   },
   "outputs": [],
   "source": [
    "options(warn=-1)\n",
    "Pkgs <- c(\"dplyr\", \"stringr\", \"WGCNA\")\n",
    "for(p in Pkgs) suppressMessages(library(p, character.only=T))\n",
    "options(stringsAsFactors = FALSE)\n",
    "# enableWGCNAThreads()\n",
    "\n",
    "Data <- read.csv(\"Result/table.tsv\", sep = \"\\t\", skip=1)\n",
    "names(Data)[1] <- \"Taxon\"\n",
    "Data <- Data %>%\n",
    "    filter(\n",
    "            str_detect(Taxon, \"g__\"),\n",
    "            !str_detect(Taxon, \"g__uncultured\"),\n",
    "            !str_detect(Taxon, \"g__Incertae\")\n",
    "    )\n",
    "rownames(Data) <- str_split_fixed(Data$Taxon, \";g__\", 2)[, 2]\n",
    "Data <- Data %>% select(-Taxon)\n",
    "Data <- t(Data)\n",
    "\n",
    "Beta <- c(1:50)\n",
    "Sft <- pickSoftThreshold(Data, powerVector=Beta)\n",
    "pdf(paste0(\"Result/WGCNA_ScaleIndependence_MG.png\"), width=960, height=480)\n",
    "    plot(\n",
    "            Sft$fitIndices[,1], -sign(Sft$fitIndices[,3])*Sft$fitIndices[,2],\n",
    "            xlab=\"Soft Threshold (power)\", ylab=\"Scale Free Topology Model Fit,signed R^2\", type=\"n\",\n",
    "            main=paste(\"Scale independence\")\n",
    "        )\n",
    "    text(\n",
    "            Sft$fitIndices[,1], -sign(Sft$fitIndices[,3])*Sft$fitIndices[,2],\n",
    "            labels=Beta, cex=0.9, col=\"red\"\n",
    "        )\n",
    "    abline(h=0.90,col=\"red\")\n",
    "dev.off()\n",
    "\n",
    "Beta <- 13\n",
    "Adj <- adjacency(Data, power=Beta);\n",
    "k <- as.vector(apply(Adj, 2, sum, na.rm=T))\n",
    "png(paste0(\"Result/WGCNA_ScaleFreePlot_MG.png\"), width=480, height=480)\n",
    "    scaleFreePlot(k, main=\"Check scale free topology\\n\")\n",
    "dev.off()\n",
    "\n",
    "TOM <- TOMsimilarity(Adj)\n",
    "DissTOM <- 1 - TOM\n",
    "GeneTree = hclust(as.dist(DissTOM), method=\"average\")\n",
    "DeepSplit <- 4\n",
    "MinModuleSize <- 10\n",
    "DynamicMods <- cutreeDynamic(\n",
    "                                dendro=GeneTree, \n",
    "                                distM=DissTOM,\n",
    "                                deepSplit=DeepSplit, \n",
    "                                pamStage=TRUE,\n",
    "                                pamRespectsDendro=TRUE,\n",
    "                                minClusterSize=MinModuleSize\n",
    "                            )\n",
    "DynamicColors <- labels2colors(DynamicMods)\n",
    "png(paste0(\"Result/WGCNA_Dendrogram_MG_1.png\"), width=960, height=480)\n",
    "    plotDendroAndColors(\n",
    "                            dendro=GeneTree, \n",
    "                            colors=DynamicColors, \n",
    "                            groupLabels=\"Dynamic Tree Cut\",\n",
    "                            dendroLabels=FALSE, hang=0.03,\n",
    "                            addGuide=TRUE, guideHang=0.05,\n",
    "                            main=\"Gene dendrogram and module colors\"\n",
    "                        )\n",
    "dev.off()\n",
    "\n",
    "MEList <- moduleEigengenes(Data, colors=DynamicColors)\n",
    "MEs <- MEList$eigengenes\n",
    "MEDiss <- 1 - cor(MEs)\n",
    "METree <- hclust(as.dist(MEDiss), method=\"average\")\n",
    "png(paste0(\"Result/WGCNA_ClusteringModule_MG.png\"), width=960, height=480)\n",
    "    plot(\n",
    "            METree,\n",
    "            main=\"Clustering of module eigengenes\",\n",
    "            xlab = \"\", sub = \"\"\n",
    "        )\n",
    "    MEDissThres <- 0.85\n",
    "    abline(h=MEDissThres, col=\"red\")\n",
    "dev.off()\n",
    "\n",
    "Merge <- mergeCloseModules(Data, DynamicColors, cutHeight=MEDissThres, verbose=3)\n",
    "MergedColours <- Merge$colors\n",
    "MergedMEs <- Merge$newMEs\n",
    "png(paste0(\"Result/WGCNA_Dendrogram_MG_2.png\"), width=960, height=480)\n",
    "    plotDendroAndColors(\n",
    "                            dendro=GeneTree,\n",
    "                            colors=cbind(DynamicColors, MergedColours),\n",
    "                            groupLabels=c(\"Dynamic Tree Cut\", \"Merged dynamic\"),\n",
    "                            dendroLabels=FALSE, hang=0.03,\n",
    "                            addGuide=TRUE, guideHang=0.05\n",
    "                        )\n",
    "dev.off()\n",
    "\n",
    "ColourList <- data.frame(colnames(Data), MergedColours)\n",
    "colnames(ColourList) <- c(\"Symbol\", \"Colour\")\n",
    "write.table(ColourList, \"Result/WGCNAR.tsv\", sep='\\t', row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abd1ca52",
   "metadata": {
    "vscode": {
     "languageId": "julia"
    }
   },
   "outputs": [],
   "source": [
    "# sankey\n",
    "using CSV, DataFrames\n",
    "using SankeyMakie\n",
    "using CairoMakie\n",
    "\n",
    "SGCRNA = CSV.read(\"Result/TaxonCluster.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "WGCNA = CSV.read(\"Result/WGCNAR.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "rename!(WGCNA, 1 => \"Taxon\");\n",
    "Data = outerjoin(SGCRNA, WGCNA, on=:Taxon);\n",
    "Data = coalesce.(Data, \"NA\");\n",
    "Data[!, :Module] = string.(Data.Module);\n",
    "gdf = groupby(Data[:,[:Module ,:Colour]], [:Module ,:Colour]);\n",
    "Data = combine(gdf, nrow);\n",
    "sort!(Data);\n",
    "Label = vcat(unique(Data.Module), unique(sort(Data.Colour)));\n",
    "\n",
    "src = [findfirst(x -> x == i, Label) for i in Data.Module];\n",
    "dst = [findfirst(x -> x == i, Label) for i in Data.Colour];\n",
    "\n",
    "cnct = [(src[i], dst[i], Data.nrow[i]) for i in 1:length(src)];\n",
    "f, ax, s = sankey(cnct, nodelabels=Label, axis=hidden_axis(), figure=(; size=(1000, 2000)))\n",
    "# hidedecorations!(ax)\n",
    "# hidespines!(ax)\n",
    "ax.fontsize = 16\n",
    "save(\"Result/MG_Sankey.png\", f)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb3d882f",
   "metadata": {},
   "source": [
    "# 要素比較ヒートマップ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e508d543",
   "metadata": {
    "vscode": {
     "languageId": "julia"
    }
   },
   "outputs": [],
   "source": [
    "using CSV, DataFrames\n",
    "using Colors, CairoMakie\n",
    "\n",
    "SGCRNA = CSV.read(\"Result/TaxonCluster.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "WGCNA = CSV.read(\"Result/WGCNAR.tsv\", header=1, comment=\"#\", delim='\\t', DataFrame);\n",
    "N_Sgcrna = maximum(SGCRNA.Module)\n",
    "ClustName = sort(unique(WGCNA.Colour))\n",
    "N_Wgcna = length(ClustName)\n",
    "Result = zeros(N_Wgcna, N_Sgcrna)\n",
    "for c in 1:N_Sgcrna\n",
    "    A = Set(SGCRNA.Taxon[SGCRNA.Module .== c])\n",
    "    for c2 in 1:N_Wgcna\n",
    "        B = Set(WGCNA.Symbol[WGCNA.Colour .== ClustName[c2]])\n",
    "        Result[c2, c] = length(A ∩ B) / min(length(A), length(B))\n",
    "    end\n",
    "end\n",
    "\n",
    "fig = Figure(size=(size(Result,1)*50, size(Result,2)*30), fontsize=16)\n",
    "ax = Axis(fig[1, 1]; xlabel=\"WGCNA\", ylabel=\"SGCRNA\", title=\"\", xgridvisible=false, ygridvisible=false, xticks=(1:size(Result,1), string.(ClustName)), xticklabelrotation=π/4, yticks=(1:size(Result,2), string.(1:size(Result,2))))\n",
    "heatmap!(ax, Result, colormap=:heat)\n",
    "Colorbar(fig[:, end+1], limits=(0, 1), colormap=:heat)\n",
    "save(\"Result/ModuleSimilarity.pdf\", fig)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scRNA_R",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "python",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
